enhancement(aws_cloudwatch_logs sink): add `json/emf` support

hencrice opened the pull request:
Made it possible to pass in arbitrary request headers such that we can
set the "json/emf" header when the logs uploaded by this sink are
actually in the Amazon CloudWatch EMF format: https://docs.aws.amazon.com/AmazonCloudWatch/latest/monitoring/CloudWatch_Embedded_Metric_Format_Specification.html#CloudWatch_Embedded_Metric_Format_Specification_structure

Closes #12760

jszwedko reviewed (approved):
Great work @hencrice ! I pushed a couple of tweaks, but this looks good to me! I like how you refactored the smithy client creation to make sure we apply the same settings (tls, proxy, etc.).

fuchsnj commented on src/sinks/aws_cloudwatch_logs/config.rs:
```diff
@@ -85,20 +90,46 @@ impl CloudwatchLogsSinkConfig {
+        &self,
+        proxy: &ProxyConfig,
+    ) -> crate::Result<aws_smithy_client::Client> {
+        let region = match self.region.region() {
```
It's probably worth breaking this out into a function in `src/aws/mod.rs` and calling it instead.

fuchsnj commented on src/aws/mod.rs:
```diff
@@ -82,6 +82,40 @@ pub trait ClientBuilder {
+    is_sink: bool,
+    retry_config: RetryConfig,
+) -> crate::Result<aws_smithy_client::Client> {
+    // Now build the client.
```
this comment probably isn't needed since it was extracted to a function

fuchsnj commented on src/sinks/aws_cloudwatch_logs/config.rs:
```diff
@@ -85,20 +90,46 @@ impl CloudwatchLogsSinkConfig {
+            .tower
+            .unwrap_with(&TowerRequestConfig::default());
         let client = self.create_client(cx.proxy()).await?;
+        let smithy_client = self.create_smithy_client(cx.proxy()).await?;
```
It would be nice if the 2 `smithy_client`s being made here could be shared, but I'm not seeing a reasonable way to do that.

fuchsnj commented on src/sinks/aws_cloudwatch_logs/service.rs:
```diff
@@ -193,9 +212,15 @@ impl Service<BatchCloudwatchRequest> for CloudwatchLogsPartitionSvc {
+        config: CloudwatchLogsSinkConfig,
         key: &CloudwatchKey,
         client: CloudwatchLogsClient,
+        smithy_client: std::sync::Arc<
```
Please create a type alias for this since it's used several times.

fuchsnj reviewed (approved):
A few small comments, but overall looks good.

hencrice commented:
@jszwedko @fuchsnj I've addressed the remaining comments. Please review.
Also a question, once this is merged, do you guys know when it will be released?

jszwedko commented:
Looks good, thanks @hencrice ! Once this is merged, it will be in the following nightly release of Vector. It will also be included in the next minor release (0.23) which is currently scheduled for July 5th.
